### ISSUE TITLE
[Enhancement] Improvement suggestion for low-level string handling

### ISSUE THREAD

**Issue by guenther-brunthaler** [wren-lang/wren#882]
I have compared the different methods how Lua and Wren handle strings.

Wren hashes the strings using a way better hash function, but unfortunately this function is also very costly, at least for long strings. It requires a multiplication per byte, which is fast on modern x86 CPUs. But not all CPUs are so fast or have hardware multiplication. Some embedded CPUs actually need a bit loop for multiplication, making Wren's current way of string hashing very slow. (Of course, they probably won't have an FPU either, but that's another topic.)

Lua on the other hand uses an extremely cheap string hashing function. But it can afford to, because the hash code is only used locally for deduplication. It is not even stored along with the string. As soon as Lua has established that only unique strings exist as Lua objects (because of the deduplication), it can use the memory address of the string as a unique identifier for it.

However, Lua's method is not necessarily faster: If only short strings are used by the program, the overhead of Wren's expensive hashing function is small, while deduplication certainly requires a more costly base effort.

On the other hand, Wren can use its expensively computed hash only for two things: For fast-detecting that two strings are *not* identical, and for using the string as a key in other hash tables.

The first use is rather uncommon - programs typically compare strings for equality and not inequality.

So the hash value is only really useful for being used as a key in another hash table, where it provides a high-quality hash value.

My improvement suggestion would now be this: Consider adopting Lua's method in the backend and deduplicate all strings in a similar way.

Then use the expensive hash function only for hashing the address of the deduplicated string, which is short and therefore not as expensive as hashing the actual string contents.

The authors of the FNV-1a hash function explicitly recommend using it for short items like the in-memory images of numeric values, and every pointer value is also just another number.

As an added bonus, low-level string comparison can be reduced to a pointer comparison, the hash is not even required for that.

And of course the memory savings could be enormous if a lot of identical strings are being processed by the script.

But deduplicated strings have more advantages than this.

In Lua, strings can also be used reliably as *atoms*!

That is, Lua does not need enum or symbols in efficient code. It can just do something like

    if state == "initial" then ...
    else if state == "blahblah" then ...
    end

because the string comparisons just involve pointer comparisons internally and are thus extremely cheap. And neither do the many repeated strings in the source file waste memory (after compiling them into internal bytecode), because they all use the same shared memory locations thanks to the deduplication.

**Comment by ruby0x1** [comment]
String literals are already interned as a constant, and string comparisons [just use the hash directly](https://github.com/wren-lang/wren/blob/main/src/vm/wren_value.c#L1316-L1319).

**Comment by mhermier** [comment]
The subject of making string `Symbol`s was a long debate with @munificent.
The concept he brought was that making `String`s constant was sufficient. I
think that the cost is not worth it at runtime, but can be interesting at
compile time.

Another solution can be to soften the hash based on hardware, or even use
processor specialized hash CPU built-ins when available.

**Comment by guenther-brunthaler** [comment]
> and string comparisons [just use the hash directly]

Yes, but the hash can only be used to determine definitive inequality.

It cannot not speed up a successful comparison for equality.

In Lua comparison for both equality and inequality is just a simple pointer comparison, no strcmp() needed, because any string with the same contents will always have the same address thanks to the deduplication.

Also note that Lua's string hashing function is lightning fast and uses even constant time, because only 5 characters of the string are included into the hash calculation at all!

While Wren's processes each and every character of the string for hashing.

Of course, the hashing is not all: Lua still needs a full string comparison after hashing, in order to rule out a hash collision. And the collision chains might be long due to Lua's weak oversimplified hash function.

So Lua needs to do more work when the string is encountered the first time AND if there are many strings in the same hash bucket.

Wren will therefore most likely beat Lua's string handling for short strings.

But if an application has to process a lot of long strings, I would assume Lua's method had to be a lot faster.

**Comment by benpop** [comment]
> Also note that Lua's string hashing function is lightning fast and uses even constant time, because only 5 characters of the string are included into the hash calculation at all!

This is not correct. For short strings, all characters are processed. If a string's length is above a certain threshold then it will skip characters, the skip width depending on the size of the string. The algorithm for that in Lua 5.1 that I can see is `step = (length >> 5) + 1;`, so if a string is less than 32 characters in length, the full string is processed.

> Of course, the hashing is not all: Lua still needs a full string comparison after hashing, in order to rule out a hash collision. And the collision chains might be long due to Lua's weak oversimplified hash function.
> 
> So Lua needs to do more work when the string is encountered the first time AND if there are many strings in the same hash bucket.

I think for at least one of these sentences you swapped "Lua" and "Wren". I'm confused by what you are saying here.

---
@mhermier Is [this the discussion](https://groups.google.com/g/wren-lang/c/F_4tI_9KaSI/m/dMQ_RyJNgZEJ) you were referencing?

**Comment by guenther-brunthaler** [comment]
> so if a string is less than 32 characters in length, the full string is processed.

Yes, you are right. Sorry. Didn't remember the code correctly.

But hashing is still constant effort for strings longer than that, and thus for all long strings where the speed of hashing eventually will become an issue.

> > Of course, the hashing is not all: Lua still needs a full string comparison after hashing, in order to rule out a hash collision. And the collision chains might be long due to Lua's weak oversimplified hash function.
> > So Lua needs to do more work when the string is encountered the first time AND if there are many strings in the same hash bucket.
> 
> I think for at least one of these sentences you swapped "Lua" and "Wren". I'm confused by what you are saying here.

What I meant was that Lua's deduplication might be slower than Wren's expensive hashing if Lua's hash function produces many hash collisions for different strings. Because then Lua needs to strcmp() the new string with all strings in the collision chain in order to see whether the new string is a duplicate of one of those strings.

The most "evil" case one can create for Lua are very long strings which have all the same size and only differ at character positions which are *not* included in the hash calculation. Therefore all those strings will yield the same hash value, yet all of them are different.

And with "when the string is encountered the first time" I meant that Lua needs to de-duplicate a new string object only once at the time of its creation. As soon as the objects exists, string comparison for both equality and inequality, whether successful or not, never needs another strcmp().

While Wren() always needs a strcmp() in order to ensure that two strings with the same hash value are indeed the same, because of the possibility of hash collisions.

Or to put it a different way: Lua does all strcmp() in advance as part of the deduplication. After that, strings can be compared with near-zero effort (just a pointer comparison) as long as they exist, and no matter how often they are used.

Wren saves the deduplication effort, but needs to call strcmp() for the same string over and over again every time it has to compare it (successfully) for equality with some other string. Lua only needs to compare pointers in this case.

**Comment by mhermier** [comment]
Yes, it is the discussion. It made me think we also needed an identity
comparison in the form of `===` and `!==` which resulted in addition of
`Ibject::same(lhs, rhs)`.
